src/music/MusicManager.ts

```diff
@@ -0,0 +1,185 @@
+import Rhythm from "./Rhythm.js";
+import { Drumkit, Drums } from "./Drums.js";
+import { Arrays, NonEmptyArray } from "../util/Arrays.js";
+
+export namespace MusicManager {
+
+    export const context: AudioContext = new AudioContext();
+
+    export const instruments = {
+        arp: new AdsrOscillatorInstrument(
+            { type: "sawtooth", detune: 3 },
+            { attack: 0.01, decay: 0.05, sustain: 0.0, release: 0.1 }
+        ),
+        bass: new AdsrOscillatorInstrument(
+            { type: "triangle", detune: 2 },
+            { attack: 0.01, decay: 0.01, sustain: 1.0, release: 0.1 }
+        ),
+        pad: new AdsrOscillatorInstrument(
+            { type: "triangle", detune: 4 },
+            { attack: 1.0, decay: 0.1, sustain: 0.5, release: 0.5 }
+        )
+    };
+
+    export const masterGain: GainNode = context.createGain();
+    export const drumkit: Drumkit = new Drumkit();
+
+    interface MusicState {
+        beatsPerMinute: number;
+        rhythm: Rhythm;
+        root: number;
+        scale: Scale;
+    }
+
+    const state: MusicState = {
+        beatsPerMinute: 220,
+        rhythm: new Rhythm(),
+        root: 60, // middle C
+        scale: 2741 // major
+    };
+
+    enum ChordFunction {
+        TONIC = "tonic",
+        SUBDOMINANT = "subdominant",
+        DOMINANT = "dominant",
+        AMBIGUOUS = "ambiguous"
+    }
+
+    const backHalves: NonEmptyArray<ChordFunction[]> = [
+        [ChordFunction.TONIC, ChordFunction.SUBDOMINANT, ChordFunction.DOMINANT],
+        [ChordFunction.SUBDOMINANT, ChordFunction.SUBDOMINANT, ChordFunction.DOMINANT],
+        [ChordFunction.SUBDOMINANT, ChordFunction.DOMINANT, ChordFunction.DOMINANT],
+        [ChordFunction.AMBIGUOUS, ChordFunction.SUBDOMINANT, ChordFunction.DOMINANT],
+        [ChordFunction.SUBDOMINANT, ChordFunction.DOMINANT, ChordFunction.AMBIGUOUS]
+    ];
+
+    const FunctionDegrees: Record<ChordFunction, NonEmptyArray<number>> = {
+        "tonic": [0, 5], // I and VI
+        "subdominant": [1, 3], // II and IV
+        "dominant": [4, 6], // V and VII
+        "ambiguous": [2] // III
+    };
+
+    function generateChordProgression(scale: Scale, root: number, extensions: number = 4): number[][] {
+        const pitchClass: number[] = Scales.getPitchClass(scale);
+        const backHalf: ChordFunction[] = Random.fromArray(backHalves);
+        // always start from the tonic and add a valid back half
+        // this way it'll always fit the tonic -> subdominant -> dominant pattern
+        const chordDegrees: number[] = [0].concat(backHalf.map(f => Random.fromArray(FunctionDegrees[f])));
+        return chordDegrees.map(degree => {
+            const chord = [];
+            for (let i = 0; i < extensions; i++) {
+                // step up scale in thirds starting from chord's root
+                chord[i] = Scales.indexIntoPitchClass(pitchClass, degree + 2 * i) + root;
+            }
+            return chord;
```

Comment:
A better way to write this would just be to modify `Arrays.generate`'s function to be able to accept an index and use that instead